stl/inc/span

```diff
@@ -0,0 +1,729 @@
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Rng, class _Ty>
+struct _Is_span_convertible_range
+    : bool_constant<is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Rng&>()))> (*)[], _Ty (*)[]>> {};
+
+// STRUCT TEMPLATE _Has_container_interface
+template <class, class = void>
+struct _Has_container_interface : false_type {};
+
+template <class _Rng>
+struct _Has_container_interface<_Rng,
+    void_t<decltype(_STD data(_STD declval<_Rng&>())), decltype(_STD size(_STD declval<_Rng&>()))>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_compatible_range
+// clang-format off
+template <class _Rng, class _Ty>
+struct _Is_span_compatible_range : bool_constant<conjunction_v<
+    negation<is_array<_Rng>>,
+    negation<_Is_span<remove_const_t<_Rng>>>,
+    negation<_Is_std_array<remove_const_t<_Rng>>>,
+    _Has_container_interface<_Rng>,
+    _Is_span_convertible_range<_Rng, _Ty>>> {};
+// clang-format on
+
+#endif // !__cpp_lib_concepts
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using size_type              = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr size_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+#ifdef __cpp_lib_concepts
+
+    // clang-format off
+    constexpr span() noexcept requires (_Extent == 0 || _Extent == dynamic_extent) = default;
+
+    template <_Is_span_compatible_iterator<element_type> _It>
+    constexpr span(_It _First, size_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_STD to_address(_First)) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        (void) (_First + _Count); // Implicit bounds checking for checked iterators
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Count == _Extent, "span construction from invalid range as count != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <_Is_span_compatible_iterator<element_type> _It, _Is_span_compatible_sentinel<_It> _Sentinel>
+    constexpr span(_It _First, _Sentinel _Last) noexcept(noexcept(_Last - _First)) // strengthened
+        : _Mybase(static_cast<size_type>(_Last - _First)), _Mydata(_STD to_address(_First)) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _Adl_verify_range(_First, _Last);

```

Comment:
Adding the changes to the other constructor it appread to me that `(static_cast<size_type>(_Last - _First)` should already perform a bounds check inside `operator-` Should we keep the `_Adl_verify_range` nevertheless?

Reply:
Hm, `_Adl_verify_range` is guarded by `_ITERATOR_DEBUG_LEVEL` wheras this constructor uses `_CONTAINER_DEBUG_LEVEL`. Does that even make sense in constext of a span?

Due to personal experience I am sadly aware that there are projects out there that handle them differently.

Reply:
We don't guard `_Adl_verify_range` at all; the design allows for iterators that are checked even in release builds (like `gsl::span`'s iterators). That said, I think you're right that we don't need the `_Adl_verify_range` here at all since we form `_Last - _First` - but let's wait for @BillyONeal to verify before making a change to avoid churn.

Reply:
Moved it out of `#if _CONTAINER_DEBUG_LEVEL > 0`